File: services/backend/app/web_evidence_provenance.py
"""Read-only trusted producer policy for Web Research Evidence."""

from __future__ import annotations

import json
import os
import re
from pathlib import Path


_SCHEMA = "web-evidence-provenance-policy.v1"
_PRODUCER_FIELDS = {
    "plugin_id", "plugin_version", "release_id", "qualification_state", "attestation_sha256"
}
_SHA256 = re.compile(r"^sha256:[0-9a-f]{64}$")


def default_policy_path() -> Path:
    configured = os.environ.get("BYQ_WEB_EVIDENCE_PROVENANCE_POLICY")
    if configured:
        return Path(configured)
    container_path = Path("/app/web-evidence-provenance.json")
    if container_path.is_file():
        return container_path
    return Path(__file__).resolve().parents[3] / "config/dsh/generated/web-evidence-provenance.json"


def load_web_evidence_provenance(path: Path | None = None) -> dict[str, object]:
    policy_path = path or default_policy_path()
    value = json.loads(policy_path.read_text(encoding="utf-8"))
    if not isinstance(value, dict) or set(value) != {
        "schema_version", "mode", "active_producer", "recognized_producers"
    }:
        raise ValueError("web evidence provenance policy has an invalid closed schema")
    if value["schema_version"] != _SCHEMA or value["mode"] not in {"qualified", "candidate"}:
        raise ValueError("web evidence provenance policy is unsupported")
    producers = value["recognized_producers"]
    if not isinstance(producers, list) or not producers:
        raise ValueError("web evidence provenance policy has no recognized producers")
    identities: set[tuple[str, str]] = set()
    for producer in producers:
        _validate_producer(producer)
        identity = (producer["plugin_id"], producer["plugin_version"])
        if identity in identities:
            raise ValueError("web evidence provenance policy has duplicate producers")
        identities.add(identity)
    active = value["active_producer"]
    _validate_producer(active)
    if active not in producers:
        raise ValueError("active web evidence producer is not recognized")
    if value["mode"] == "qualified" and any(
        item["qualification_state"] != "QUALIFIED" for item in producers
    ):
        raise ValueError("qualified web evidence policy contains an unqualified producer")
    if value["mode"] == "qualified" and active["qualification_state"] != "QUALIFIED":
        raise ValueError("default web evidence producer is not qualified")
    if value["mode"] == "candidate" and active["qualification_state"] != "CANDIDATE":
        raise ValueError("candidate web evidence policy does not identify a candidate")
    return value


def _validate_producer(value: object) -> None:
    if not isinstance(value, dict) or set(value) != _PRODUCER_FIELDS:
        raise ValueError("web evidence producer has an invalid closed schema")
    if value["plugin_id"] != "web-search":
        raise ValueError("web evidence producer plugin is unsupported")
    for field in ("plugin_version", "release_id", "qualification_state"):
        if not isinstance(value[field], str) or not value[field]:
            raise ValueError(f"web evidence producer {field} is invalid")
    if value["qualification_state"] not in {"QUALIFIED", "CANDIDATE"}:
        raise ValueError("web evidence producer qualification state is invalid")
    if not isinstance(value["attestation_sha256"], str) or _SHA256.fullmatch(value["attestation_sha256"]) is None:
        raise ValueError("web evidence producer attestation is invalid")


def recognized_producer(plugin_id: object, plugin_version: object) -> bool:
    policy = load_web_evidence_provenance()
    return any(
        item["plugin_id"] == plugin_id and item["plugin_version"] == plugin_version
        for item in policy["recognized_producers"]
    )

File: services/mcp/src/web-evidence-provenance.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

export type WebEvidenceProducer = {
  plugin_id: "web-search";
  plugin_version: string;
  release_id: string;
  qualification_state: "QUALIFIED" | "CANDIDATE";
  attestation_sha256: string;
};

type WebEvidencePolicy = {
  schema_version: "web-evidence-provenance-policy.v1";
  mode: "qualified" | "candidate";
  active_producer: WebEvidenceProducer;
  recognized_producers: WebEvidenceProducer[];
};

const PRODUCER_FIELDS = [
  "attestation_sha256", "plugin_id", "plugin_version", "qualification_state", "release_id",
].sort();

function policyPath(): string {
  return process.env.BYQ_WEB_EVIDENCE_PROVENANCE_POLICY
    ?? (process.cwd() === "/app"
      ? "/app/web-evidence-provenance.json"
      : resolve(process.cwd(), "../../config/dsh/generated/web-evidence-provenance.json"));
}

function producer(value: unknown): WebEvidenceProducer {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("web evidence producer must be an object");
  }
  const item = value as Record<string, unknown>;
  if (JSON.stringify(Object.keys(item).sort()) !== JSON.stringify(PRODUCER_FIELDS)
      || item.plugin_id !== "web-search"
      || typeof item.plugin_version !== "string" || !item.plugin_version
      || typeof item.release_id !== "string" || !item.release_id
      || !["QUALIFIED", "CANDIDATE"].includes(String(item.qualification_state))
      || typeof item.attestation_sha256 !== "string"
      || !/^sha256:[0-9a-f]{64}$/.test(item.attestation_sha256)) {
    throw new Error("web evidence producer is invalid");
  }
  return item as WebEvidenceProducer;
}

export function loadWebEvidencePolicy(): WebEvidencePolicy {
  const value = JSON.parse(readFileSync(policyPath(), "utf8")) as Record<string, unknown>;
  if (JSON.stringify(Object.keys(value).sort()) !== JSON.stringify([
    "active_producer", "mode", "recognized_producers", "schema_version",
  ]) || value.schema_version !== "web-evidence-provenance-policy.v1"
      || !["qualified", "candidate"].includes(String(value.mode))
      || !Array.isArray(value.recognized_producers) || value.recognized_producers.length === 0) {
    throw new Error("web evidence provenance policy is invalid");
  }
  const active = producer(value.active_producer);
  const recognized = value.recognized_producers.map(producer);
  const identities = new Set(recognized.map((item) => `${item.plugin_id}\u0000${item.plugin_version}`));
  if (identities.size !== recognized.length || !recognized.some((item) => PRODUCER_FIELDS.every((field) => item[field as keyof WebEvidenceProducer] === active[field as keyof WebEvidenceProducer]))) {
    throw new Error("web evidence provenance policy producer set is invalid");
  }
  if (value.mode === "qualified" && recognized.some((item) => item.qualification_state !== "QUALIFIED")) {
    throw new Error("qualified web evidence policy contains an unqualified producer");
  }
  if ((value.mode === "qualified" && active.qualification_state !== "QUALIFIED")
      || (value.mode === "candidate" && active.qualification_state !== "CANDIDATE")) {
    throw new Error("web evidence provenance policy mode is inconsistent");
  }
  return { schema_version: value.schema_version, mode: value.mode, active_producer: active, recognized_producers: recognized } as WebEvidencePolicy;
}

export function bindActiveWebEvidenceProducer(content: Record<string, unknown>): Record<string, unknown> {
  const search = content.search;
  if (search === null || typeof search !== "object" || Array.isArray(search)) return content;
  const active = loadWebEvidencePolicy().active_producer;
  const supplied = search as Record<string, unknown>;
  for (const field of ["plugin_id", "plugin_version"] as const) {
    if (field in supplied && supplied[field] !== active[field]) {
      throw new Error("web evidence producer claim does not match trusted deployment");
    }
  }
  return {
    ...content,
    search: { ...supplied, plugin_id: active.plugin_id, plugin_version: active.plugin_version },
  };
}
